drivers/spi/spi_mchp_mss.c

```diff
@@ -0,0 +1,496 @@
+
+	if (mss_spi_clk_gen_set(cfg, spi_cfg)) {
+		LOG_ERR("can't set clk divider\n");
+		return -EINVAL;
+	}
+
+	mss_spi_hw_mode_set(cfg, spi_cfg->operation);
+
+	mss_spi_write(cfg, MSS_SPI_REG_TXRXDF_SIZE, MSS_SPI_FRAMESIZE_DEFAULT);
+	control = mss_spi_read(cfg, MSS_SPI_REG_CONTROL);
+	control |= MSS_SPI_CONTROL_ENABLE;
+	mss_spi_write(cfg, MSS_SPI_REG_CONTROL, control);
+	if (spi_cfg->operation & (SPI_TRANSFER_LSB | SPI_OP_MODE_SLAVE | SPI_MODE_LOOP)) {
+		LOG_WRN("not supported operation\n\r");
+		return -ENOTSUP;
+	}
+
+	mss_spi_write(cfg, MSS_SPI_REG_COMMAND, 0xC);
+
+	return 0;
+}
+
+static int mss_spi_transceive(const struct device *dev, const struct spi_config *spi_cfg,
+			      const struct spi_buf_set *tx_bufs, const struct spi_buf_set *rx_bufs,
+			      bool async, spi_callback_t cb, void *userdata)
+{
+
+	const struct mss_spi_config *config = dev->config;
+	struct mss_spi_data *data = dev->data;
+	struct spi_context *ctx = &data->ctx;
+	struct mss_spi_transfer *xfer = &data->xfer;
+	int ret = 0;
+
+	spi_context_lock(ctx, async, cb, userdata, spi_cfg);
+
+	ret = mss_spi_configure(dev, spi_cfg);
+	if (ret) {
+		LOG_ERR("Fail to configure\n\r");
+		goto out;
+	}
+
+	ctx->config = spi_cfg;
+	spi_context_buffers_setup(ctx, tx_bufs, rx_bufs, 1);
+	xfer->rx_len = ctx->rx_len;
+	mss_spi_readwr_fifo(dev);
+	ret = spi_context_wait_for_completion(ctx);
+out:
+	spi_context_release(ctx, ret);
+	mss_spi_disable_ints(config);
+
+	return ret;
+}
+
+static int mss_spi_transceive_blocking(const struct device *dev, const struct spi_config *spi_cfg,
+				       const struct spi_buf_set *tx_bufs,
+				       const struct spi_buf_set *rx_bufs)
+{
+
+	return mss_spi_transceive(dev, spi_cfg, tx_bufs, rx_bufs, false, NULL, NULL);
+}
+
+#ifdef CONFIG_SPI_ASYNC
+static int mss_spi_transceive_async(const struct device *dev, const struct spi_config *spi_cfg,
+				    const struct spi_buf_set *tx_bufs,
+				    const struct spi_buf_set *rx_bufs, spi_callback_t cb,
+				    void *userdata)
+{
+	return mss_spi_transceive(dev, spi_cfg, tx_bufs, rx_bufs, true, cb, userdata);
+}
+#endif /* CONFIG_SPI_ASYNC */
+
+static int mss_spi_init(const struct device *dev)
+{
+	const struct mss_spi_config *cfg = dev->config;
+	struct mss_spi_data *data = dev->data;
+	struct mss_spi_transfer *xfer = &data->xfer;
+	int ret = 0;
+	uint32_t control = 0;
+
+	/* Remove SPI from Reset  */
+	control = mss_spi_read(cfg, MSS_SPI_REG_CONTROL);
+	control &= ~MSS_SPI_CONTROL_RESET;
+	mss_spi_write(cfg, MSS_SPI_REG_CONTROL, control);
+
+	/* Set master mode */
+	control = mss_spi_read(cfg, MSS_SPI_REG_CONTROL);
+	control &= ~MSS_SPI_CONTROL_ENABLE;
+	mss_spi_write(cfg, MSS_SPI_REG_CONTROL, control);
+
+	xfer->control = (MSS_SPI_CONTROL_SPS | MSS_SPI_CONTROL_BIGFIFO | MSS_SPI_CONTROL_MASTER |
+			 MSS_SPI_CONTROL_CLKMODE);
+
+	control = mss_spi_read(cfg, MSS_SPI_REG_CONTROL);
+	control |= MSS_SPI_CONTROL_ENABLE;
+	mss_spi_write(cfg, MSS_SPI_REG_CONTROL, control);
```

Comment:
> I don't know anything about this controller, but you seem to disable/enable the controller without doing anything to it. xfer->control is not written anywhere.
> 
> Btw, are you sure that the controller must stay enabled while not transferring? Should it be enabled before configuring, and disabled in case of error or at the end of the transaction?

oh, understood. It should be enabled at the start of transfer and disable after completion of transfer.
Let me recheck the configuration, while developing the driver I have seen that with out disabling the core, I can't be able to set any of the bits in the controller register. But as mentioned by you, let me re frame the controller enable/disable task.

Reply:
I missed that, but beware there is a difference with enabled to transfer and enabled about power.  If that's about power, just leave it out.

Reply:
Hi @tbursztyka,

Sorry for the late reply. its about enabled to transfer. as you mentioned previously, I reframed the controller enable/disable task. In the initial version of the driver the controller enable/disable is taken place at many places. but in this version, as you said it is enabled before configuration and disabled at the end of transfer.

Reply:
Hi @tbursztyka, could I ask you to confirm your changes have been addressed so that we can get this PR merged?
Thanks a million!